ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs

```diff
@@ -0,0 +1,257 @@
+                memberDeclaration.NameIdentifier.GetDocumentRange().EndOffsetRange()
+
+            | _ -> equalsToken.GetDocumentRange().StartOffsetRange()
+
+        let symbolUse =
+            match decl with
+            | :? IBinding as binding ->
+                let refPat = binding.HeadPattern.As<IReferencePat>()
+                if isNull refPat then Unchecked.defaultof<_> else
+                refPat.GetFcsSymbolUse()
+            | :? IMemberDeclaration as memberDeclaration ->
+                memberDeclaration.GetFcsSymbolUse()
+            | _ -> Unchecked.defaultof<_>
+
+        if isNull symbolUse then ValueNone else
+        let symbol = symbolUse.Symbol.As<FSharpMemberOrFunctionOrValue>()
+        if isNull symbol then ValueNone else
+
+        let displayContext = symbolUse.DisplayContext
+        createTypeHintHighlighting symbol.ReturnParameter.Type displayContext range pushToHintMode |> ValueSome
+
+    let rec getHintForPattern (pattern: IFSharpPattern) pushToHintMode =
+        match pattern with
+        | :? IParametersOwnerPat as pattern ->
+            let asPat = AsPatNavigator.GetByLeftPattern(pattern.IgnoreParentParens())
+            if isNull asPat then ValueNone else
+            let reference = pattern.Reference
+            if isNull reference then ValueNone else
+            let symbol = reference.GetFcsSymbol().As<FSharpActivePatternCase>()
+            if isNull symbol then ValueNone
+            else getHintForPattern (asPat.RightPattern.IgnoreInnerParens()) pushToHintMode
+
+        | :? IReferencePat as refPat ->
+            let symbolUse = refPat.GetFcsSymbolUse()
+            if isNull symbolUse then ValueNone else
+            let symbol = symbolUse.Symbol.As<FSharpMemberOrFunctionOrValue>()
+            if isNull symbol then ValueNone else
+
+            let fcsType = symbol.FullType
+            let range = pattern.GetNavigationRange().EndOffsetRange()
+            let displayContext = symbolUse.DisplayContext
+
+            createTypeHintHighlighting fcsType displayContext range pushToHintMode |> ValueSome
+
+        | _ -> ValueNone
+
+    let rec getHighlighting (node: ITreeNode) pushToHintMode =
+        match node with
+        | :? IFSharpPattern as pattern ->
+            getHintForPattern pattern pushToHintMode
+
+        | :? IParameterOwnerMemberDeclaration as decl ->
+            getReturnTypeHint decl pushToHintMode
+
+        | :? IAutoPropertyDeclaration as decl ->
+            let symbolUse = decl.GetFcsSymbolUse()
+            if isNull symbolUse then ValueNone else
+            let symbol = symbolUse.Symbol.As<FSharpMemberOrFunctionOrValue>()
+            if isNull symbol then ValueNone else
+
+            let fcsType = symbol.ReturnParameter.Type
+            let range = decl.NameIdentifier.GetNavigationRange().EndOffsetRange()
+            let displayContext = symbolUse.DisplayContext.WithShortTypeNames(true)
```

Comment:
Parts of this code seem to be repeated here and there. Could you try to extract it and reuse?

Reply:
It's good that you noticed. In general, this entire piece of code should be deleted, since we don't show hints for auto-properties (just like we didn't show Code Vision)